File: 40-combination-sum-ii/combination-sum-ii.java
class Solution {

    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        List<List<Integer>> result = new ArrayList<>();

        Arrays.sort(candidates);

        explore(candidates, result, new ArrayList<>(), target, 0);

        return result;
    }

    private void explore(int[] candidates, List<List<Integer>> result, List<Integer> curr, int target, int i) {
       if (target == 0) {
        result.add(new ArrayList<>(curr));
        return;
       }
       if (target < 0 || i >= candidates.length) return ;
        for (int j = i; j < candidates.length; j++) {
            if (j > i && candidates[j] == candidates[j-1]) {
                continue;
            }

            curr.add(candidates[j]);
            explore(candidates, result, curr, target - candidates[j], j + 1);
            curr.remove(curr.size() - 1);
        }
    }
}

File: 40-combination-sum-ii/combination-sum-ii.py
# https://leetcode.com/problems/combination-sum-ii/

from typing import List

class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        results=[]

        def solve(arr, i, current, ssf, target):
            nonlocal results
            if ssf == target:
                results.append(current[:])
                return 
            if i == len(arr) or ssf > target:
                return 0
            
            n=len(arr)
            for j in range(i, n):
                if j>i and arr[j] == arr[j-1]:
                    continue
                current.append(arr[j])
                f=solve(arr,j+1,current,ssf + arr[j], target)
                current.pop()

        # we need to sort the candidates as it is asked to 
        # use the numbers only once.
        candidates.sort()

        solve(candidates,0,[],0,target)
        return results
